### PR TITLE
fix: replace I_MPI_SUBSTITUTE_INSTALLDIR with actual installation prefix in mpicc and related scripts

### PR DESCRIPTION
**Pull request by tvandera**
If you do not do this `mpicc -compile-info` will contain flags with `I_MPI_SUBSTITUTE_INSTALLDIR` instead of the actual path.

### CODE REVIEW

**Comment by alalazo** [comment]
@danvev  @rscohn2

**Comment by rscohn2** [comment]
Thanks for the PR. Can you give more details on how you using the package? It works fine if you do a
```
spack load intel-oneapi-mpi
```

Are you doing:
```
spack location -i intel-oneapi-mpi
```
To get the location and then invoking it?

**Comment by tvandera** [comment]
This seems to be an issue when you build a package that depends on `intel-oneapi-mpi` as opposed to `spack load intel-oneapi-mpi`. When you `spack load intel-oneapi-mpi`, the `I_MPI_ROOT` environment variable is set, and hides the `_MPI_SUBSTITUTE_INSTALLDIR` issue in mpicc:

This is the relevant part from `mpigcc`:
```
# Directory locations: Fixed for any MPI implementation.
# Set from the directory arguments to configure (e.g., --prefix=/usr/local)
prefix=I_MPI_SUBSTITUTE_INSTALLDIR
# The environment variable I_MPI_ROOT may be used to override installation folder path
if [ -n "$I_MPI_ROOT" ] ; then
    prefix=$I_MPI_ROOT;
fi
```

but when you build a dependent package this does not happen.

```
[/tmp/vanderaa/spack-stage/spack-stage-parallel-netcdf-1.12.2-ybxkgwlbf2bfmbajmfskulrql5axqz6u]
@vanderaa@lynx$ grep MPI spack-build-env.txt
MPICC=/home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdangcl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpicc; export MPICC
MPICH_CC=/home/vanderaa/epeec/spack/spack/lib/spack/env/gcc/gcc; export MPICH_CC
MPICH_CXX=/home/vanderaa/epeec/spack/spack/lib/spack/env/gcc/g++; export MPICH_CXX
MPICH_F77=/home/vanderaa/epeec/spack/spack/lib/spack/env/gcc/gfortran; export MPICH_F77
MPICH_F90=/home/vanderaa/epeec/spack/spack/lib/spack/env/gcc/gfortran; export MPICH_F90
MPICH_FC=/home/vanderaa/epeec/spack/spack/lib/spack/env/gcc/gfortran; export MPICH_FC
MPICXX=/home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdangcl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpicxx; export MPICXX
MPIF77=/home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdangcl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpif77; export MPIF77
MPIF90=/home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdangcl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpif90; export MPIF90
MPIFC=/home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdangcl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpifc; export MPIFC
SPACK_COMPILER_SPEC=gcc@9.3.0; export SPACK_COMPILER_SPEC
```
 Hence the error:
```
     815    make[2]: Entering directory '/tmp/vanderaa/spack-stage/spack-stage-parallel-netcdf-1.12.2-ybxkgwlbf2bfmbajmfskulrql5axqz6u/spack-src/src/libs'
     816    /bin/bash ../../libtool  --tag=CXX   --mode=link /home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/intel-oneapi-mpi-2021.2.0-aezidhdang
            cl57a6qtyrwwgriuznkzid/mpi/2021.2.0/bin/mpicxx  -fPIC -version-info 4:2:0  -o libpnetcdf.la -rpath /home/vanderaa/epeec/spack/spack/opt/spack/linux-ubuntu20.04-westm
            ere/gcc-9.3.0/parallel-netcdf-1.12.2-ybxkgwlbf2bfmbajmfskulrql5axqz6u/lib   ../dispatchers/libdispatchers.la ../drivers/common/libcommon.la ../drivers/ncmpio/libncmp
            io.la     ../binding/cxx/libcxx.la ../binding/f77/libf77.la ../binding/f90/libf90.la -LI_MPI_SUBSTITUTE_INSTALLDIR/lib/release -LI_MPI_SUBSTITUTE_INSTALLDIR/lib -L/u
            sr/lib/gcc/x86_64-linux-gnu/9 -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib -L/lib/x86_64-linux-gnu -
            L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/9/../../.. -lmpifort -lmpi -lrt -lpthread -ldl -lgfortran -lm -lquadmath -
            LI_MPI_SUBSTITUTE_INSTALLDIR/lib/release -LI_MPI_SUBSTITUTE_INSTALLDIR/lib -L/usr/lib/gcc/x86_64-linux-gnu/9 -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-
            gnu -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux
            -gnu/9/../../.. -lmpifort -lmpi -lrt -lpthread -ldl -lgfortran -lm -lquadmath
  >> 817    ../../libtool: line 7537: cd: I_MPI_SUBSTITUTE_INSTALLDIR/lib/release: No such file or directory
```

**Review by rscohn2 — APPROVED** [review]
Looks good. Thanks!

**Comment by rscohn2** [comment]
@DanVev: I think this is a bug in the intel installer, can you report it? The mpigcc and related scripts need to be patched with the paths after the install location is known. This PR makes spack do the patching.

Testing is probably not showing the problem because they are sourcing the setvars.sh, but people typically use mpigcc/mpicc scripts so they can get all the necessary compile/link options without manually adding compile options or setting environment variables.

**Comment by vdanilin** [comment]
@rscohn2 I'll create a task and check with the MPI product team.

**Comment by vdanilin** [comment]
@rscohn2 the issue has been reported and I was said out that these variables will be removed from MPI scripts in next releases.

**Comment by rscohn2** [comment]
> @rscohn2 the issue has been reported and I was said out that these variables will be removed from MPI scripts in next releases.

Will it be required that I_MPI_ROOT be defined to use mpigcc? It would be better if the installer patched the install root into the script so no environment variables are needed.

**Comment by vdanilin** [comment]
@rscohn2 we used to patch these variables on the installer side but this logic has been removed by MPI request. As I understand, they want to remove any path hardcode from the code.

**Comment by rscohn2** [comment]
@alalazo: Can you enable testing for this PR? I believe it is ready to merge assuming all tests pass. The root cause is a bug in the Intel installer, and this is a good workaround to make the package usable.

**Review by alalazo — APPROVED** [review]
Already approved by maintainer

**Pull request merged**
